internal/graph/check.go

```diff
@@ -0,0 +1,485 @@
+		g := new(errgroup.Group)
+
+		g.Go(func() error {
+
+			tuple, err := cc.datastore.ReadUserTuple(ctx, req.WrappedRequest.StoreId, req.WrappedRequest.TupleKey)
+			if err != nil {
+				if errors.Is(err, storage.ErrNotFound) {
+					resultChan <- &dispatchpb.DispatchCheckResponse{
+						Metadata: emptyMetadata,
+						WrappedResponse: &openfgapb.CheckResponse{
+							Allowed: false,
+						},
+					}
+
+					return nil
+				}
+
+				return err
+			}
+
+			_ = tuple // todo: append the resolution path to the tracer
+
+			resultChan <- &dispatchpb.DispatchCheckResponse{
+				WrappedResponse: &openfgapb.CheckResponse{
+					Allowed: true,
+				},
+			}
+
+			return nil
+		})
+
+		g.Go(func() error {
+			iter, err := cc.datastore.ReadUsersetTuples(ctx, req.WrappedRequest.StoreId, req.WrappedRequest.TupleKey)
+			if err != nil {
+				return err
+			}
+
+			var requestsToDispatch []ReduceableCheckFunc
+			for {
+				t, err := iter.Next()
+				if err != nil {
+					if err == storage.TupleIteratorDone {
+						break
+					}
+					return err
+				}
+
+				tupleKey := t.GetKey()
+				object, relation := tuple.SplitObjectRelation(tupleKey.GetUser())
+
+				requestsToDispatch = append(
+					requestsToDispatch,
+					cc.dispatch(&dispatchpb.DispatchCheckRequest{
+						Metadata: decrementDepth(req.GetMetadata()),
+						WrappedRequest: &openfgapb.CheckRequest{
+							StoreId: req.GetWrappedRequest().GetStoreId(),
+							TupleKey: &openfgapb.TupleKey{
+								Object:   object,
+								Relation: relation,
+								User:     req.GetWrappedRequest().GetTupleKey().GetUser(),
+							},
+						},
+					}),
+				)
+			}
+			iter.Stop()
+
+			resolved, err := any(ctx, requestsToDispatch)
+			if err != nil {
+				return err
+			}
+
+			resultChan <- resolved
+			return nil
+		})
+
+		return g.Wait()
+
+	}
+}
+
+func (cc *ConcurrentChecker) checkComputedUserset(ctx context.Context, req *dispatchpb.DispatchCheckRequest, rw *openfgapb.Userset_ComputedUserset) ReduceableCheckFunc {
+	return cc.dispatch(&dispatchpb.DispatchCheckRequest{
+		Metadata: decrementDepth(req.GetMetadata()),
+		WrappedRequest: &openfgapb.CheckRequest{
+			StoreId: req.GetWrappedRequest().GetStoreId(),
+			TupleKey: &openfgapb.TupleKey{
+				Object:   req.GetWrappedRequest().GetTupleKey().GetObject(),
+				Relation: rw.ComputedUserset.GetObject(),
+				User:     req.GetWrappedRequest().GetTupleKey().GetUser(),
+			},
+		},
+	})
+}
+
+func (cc *ConcurrentChecker) checkTupleToUserset(
+	ctx context.Context,
+	req *dispatchpb.DispatchCheckRequest,
+	rw *openfgapb.Userset_TupleToUserset,
+) ReduceableCheckFunc {
+	return func(ctx context.Context, resultChan chan<- *dispatchpb.DispatchCheckResponse) error {
+
+		relation := rw.TupleToUserset.GetTupleset().GetRelation()
+		if relation == "" {
+			relation = req.GetWrappedRequest().GetTupleKey().GetRelation()
+		}
+
+		var requestsToDispatch []ReduceableCheckFunc
+		for {
+			iter, err := cc.datastore.Read(
```

Comment:
Our storage interface could also use a little rework in this regard as well. It would be nice if our storage interface took optional filter parameters that you could provide to filter by the different filters that you may want instead of passing in a `TupleKey` with omitted fields to get the filter you desire. This behavior of the storage interface is nuanced and unclear and it would make our code read easier if there was a single read method such as 

```
type RelationshipFilter struct {
    ObjectType string
    OptionalObjectID string
    OptionalRelation string
    OptionalUserFilter string
}

func QueryRelationships(
    ctx context.Context,
    filter *RelationshipFilter,
    opts ...QueryOptionsOption,
) (*RelationshipIterator, error) {...}
```